src/cli/playbook/input.rs

```diff
@@ -0,0 +1,154 @@
+use crate::ff::Field;
+
+use std::any::type_name;
+use std::fs::File;
+use std::io;
+use std::io::{stdin, BufRead, BufReader, Read};
+
+use std::path::PathBuf;
+
+pub trait InputItem {
+    fn from_str(s: &str) -> Self;
+}
+
+impl<F: Field> InputItem for F {
+    fn from_str(s: &str) -> Self {
+        let int_v = s.parse::<u128>().unwrap();
+        F::from(int_v)
+    }
+}
+
+impl InputItem for u64 {
+    fn from_str(s: &str) -> Self {
+        s.parse::<u64>().unwrap()
+    }
+}
+
+impl<I: InputItem> InputItem for (I, I) {
+    fn from_str(s: &str) -> Self {
+        let mut iter = s.split(',');
+        match (iter.next(), iter.next()) {
+            (Some(left), Some(right)) => (I::from_str(left), I::from_str(right)),
+            _ => panic!(
+                "{s} is not a valid tuple of input elements: {}",
+                type_name::<I>()
+            ),
+        }
+    }
+}
+
+pub struct InputSource {
+    inner: Box<dyn BufRead>,
+}
+
+impl InputSource {
+    /// Opens a new input source from the given file.
+    ///
+    /// ## Panics
+    /// This function will panic  if `path` does not already exist.
+    #[must_use]
+    pub fn from_file(path: &PathBuf) -> Self {
+        Self {
+            inner: Box::new(BufReader::new(File::open(path).unwrap())),
+        }
+    }
+
+    #[must_use]
+    pub fn from_stdin() -> Self {
+        Self {
+            inner: Box::new(BufReader::new(stdin())),
+        }
+    }
+
+    #[cfg(test)]
+    #[must_use]
+    pub fn from_static_str(input: &'static str) -> Self {
+        Self {
+            inner: Box::new(BufReader::new(input.as_bytes())),
+        }
+    }
+
+    pub fn iter<T: InputItem>(self) -> impl Iterator<Item = T> {
+        self.lines()
+            .filter_map(|line| line.map(|l| T::from_str(&l)).ok())
+    }
+}
+
+impl Read for InputSource {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        self.inner.read(buf)
+    }
+}
+
+impl BufRead for InputSource {
+    fn fill_buf(&mut self) -> io::Result<&[u8]> {
+        self.inner.fill_buf()
+    }
+
+    fn consume(&mut self, amt: usize) {
+        self.inner.consume(amt);
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::cli::playbook::input::InputItem;
+    use crate::ff::{Fp31, Fp32BitPrime};
+    use crate::secret_sharing::IntoShares;
+    use crate::test_fixture::Reconstruct;
+
+    #[test]
+    fn from_str() {
+        assert_eq!(Fp31::from(1_u128), Fp31::from_str("1"));
+        assert_eq!(Fp32BitPrime::from(0_u128), Fp32BitPrime::from_str("0"));
+        assert_eq!(6_u64, u64::from_str("6"));
+    }
+
+    #[test]
+    #[should_panic]
+    fn parse_negative() {
+        Fp31::from_str("-1");
+    }
+
+    #[test]
+    #[should_panic]
+    fn parse_empty() {
+        Fp31::from_str("");
+    }
+
+    #[test]
+    fn tuple() {
+        let input = "20,27";
+        let tp = <(Fp31, Fp31)>::from_str(input);
+        let shares = tp.share();
```

Comment:
I wanted to make sure I can call `.share` on `TestInput` because that is what I would need later when parsing the input on the client side. `reconstruct` is because I did not want to validate individual shares :)

Reply:
👍